oops/backplane/gold_master.py

```diff
@@ -0,0 +1,2112 @@
+
+    def save_browse(self, array, browse_path):
+        """Save a backplane as a PNG, JPG, or TIFF file."""
+
+        # Get pixels and mask
+        if isinstance(array, np.ndarray):
+            image = array.copy()
+            mask = False
+        else:
+            image = array.vals.copy()
+            mask = array.mask
+
+        image = np.asfarray(image)
+        image = np.atleast_2d(image)
+
+        # Handle the mask
+        if np.all(mask):        # if fully masked
+            image.fill(0.)
+            minval = 0.
+            maxval = 1.
+
+        elif array.dtype() == 'bool':
+            # masked -> -0.2 (black); False -> 0 (dark gray); True -> 1 (white)
+            minval = -0.2
+            maxval = 1.
+            image = image.astype('int')
+            image[mask] = minval
+
+        elif np.any(mask):
+            # masked is black; minimum unmasked value is very dark gray
+            minval = np.min(image[np.logical_not(mask)])
+            maxval = np.max(image[np.logical_not(mask)])
+            new_minval = minval - 0.05 * (maxval - minval)
+            image[mask] = new_minval
+            minval = new_minval
+
+        else:
+            minval = np.min(image)
+            maxval = np.max(image)
+
+        # Create the scaled array of bytes
+        if minval == maxval:
+            scaled_bytes = np.zeros(image.shape, dtype=np.int8)
+        else:
+            scaled_floats = (image - minval) / float(maxval - minval)
+            scaled_bytes = (255.999999 * scaled_floats).astype('uint8')
+
+        # Apply zoom
+        if self.args.zoom != 1:
+            scaled_bytes = zoom_image(scaled_bytes, self.zoom, order=0)
+
+        # Make sure y-axis increases downward
+        if self.upward:
+            scaled_bytes = scaled_bytes[::-1]
+
+        shape = scaled_bytes.shape[::-1]
+        im = PIL.Image.frombytes('L', shape, scaled_bytes)
+        im.save(browse_path)
+
+    #===========================================================================
+    @staticmethod
+    def read_browse(browse_path):
+        """Read a PNG, JPG, or TIFF image file as a 2-D array of unsigned bytes.
+        """
+
+        with PIL.Image.open(browse_path, mode='r') as im:
+            return np.array(im)
+
+    ############################################################################
+    # Result summary support
+    #
+    # The summary file is a text file containing the definition of a Python
+    # dictionary. The dictionary is keyed by backplane title string and its
+    # values are tuples (minimum value, maximum value, masked pixels, total
+    # pixels). If the minimum and maximum are are equal, only one value is
+    # listed; if the object is fully masked, the value is None.
+    ############################################################################
+
+    @property
+    def gold_summary(self):
+
+        if self.gold_summary_ is not None:
```

Comment:
Yeah, I use it for trailing slashes sometimes too. The issue here is that "gold_summary" is the property, so the attribute needs a different name, and the trailing underscore fits the bill. It's similar to naming a list "list_" or a tuple "tuple_", because you want to avoid conflicting with a reserved word. The leading underscore is a bit different--to distinguish "public" from "quasi-private" attributes, but that's not my meaning here---why would we need to bother distinguishing between the public and private attributes of the BackplaneTest class, when for all practical purposes they are all private. In PdsFile I would have call this attribute something like "gold_summary_filled", but that started to seem like a lot of extraneous typing.